### PR TITLE
executor: inject random panic to AggExec

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary: 
In the project Tracking_Aggregate_Memory, there are many tracker.Consume, that will trigger panic when memory usage is too large. We should keep TiDB can cancel the sql successfully, and doesn't leak some resource.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed: 

How it Works:

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: executor/aggregate.go [hunk 1]
```diff
@@ -293,7 +294,7 @@ func (e *HashAggExec) initForParallelExec(ctx sessionctx.Context) {
 	finalConcurrency := sessionVars.HashAggFinalConcurrency()
 	partialConcurrency := sessionVars.HashAggPartialConcurrency()
 	e.isChildReturnEmpty = true
-	e.finalOutputCh = make(chan *AfFinalResult, finalConcurrency)
+	e.finalOutputCh = make(chan *AfFinalResult, finalConcurrency+partialConcurrency+1)
```
Comment (XuHuaiyu) [inline executor/aggregate.go:297]: Why change this?

File: executor/aggregate.go [hunk 2]
```diff
@@ -736,8 +743,9 @@ func (e *HashAggExec) waitPartialWorkerAndCloseOutputChs(waitGroup *sync.WaitGro
 	}
 }
 
-func (e *HashAggExec) waitFinalWorkerAndCloseFinalOutput(waitGroup *sync.WaitGroup) {
+func (e *HashAggExec) waitAllWorkerAndCloseFinalOutput(waitGroup *sync.WaitGroup, waitGroup2 *sync.WaitGroup) {
 	waitGroup.Wait()
+	waitGroup2.Wait()
```
Comment (XuHuaiyu) [inline executor/aggregate.go:748]: Why change this?

File: executor/aggregate_test.go [hunk 3]
```diff
@@ -1297,3 +1301,55 @@ func (s *testSuiteAgg) TestIssue20658(c *C) {
...
+	tk.MustExec("set @@tidb_init_chunk_size=1")
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int)")
+	for i := 0; i <= 1000; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values(%v),(%v),(%v)", i, i, i))
+	}
+
+	fpName := "github.com/pingcap/tidb/executor/ConsumeRandomPanic"
+	c.Assert(failpoint.Enable(fpName, `1%panic("ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]")`), IsNil)
```
Comment (XuHuaiyu) [inline executor/aggregate_test.go:1317]: Why is this a **random** panic?

File: executor/aggregate.go (hunk 1 above)
Comment (wshwsh12) [reply executor/aggregate.go]: revert.

File: executor/aggregate.go (hunk 2 above)
Comment (wshwsh12) [reply executor/aggregate.go]: revert.

File: executor/aggregate_test.go (hunk 3 above)
Comment (wshwsh12) [reply executor/aggregate_test.go:1317]: I use `1%` to control the panic frequency.

File: executor/aggregate.go (hunk 2 above)
Comment (wshwsh12) [reply executor/aggregate.go]: When there causes any error, HashAggExec will close finishCh, and all goroutine will try to exit.  We can't guarantee fetchChildWorker and partialWorker will exit before finalWorker.
If all finalWorkers exit first and close finalOutputCh, recoveryHashAgg() in fetchChildWorker/partialWorker will try to send message to a closed chan if there are some panic..

In my random panic test, I find the data race.

File: executor/aggregate.go (hunk 1 above)
Comment (wshwsh12) [reply executor/aggregate.go:297]: Not only finalWorker will send data/error to the channel, but also fetchChildWorker/partialWorker will send errors to the channel.
If the buffer is not enough, the send will be blocked, and maybe lead dead lock.

File: executor/aggregate_test.go (hunk 3 above)
Comment (wshwsh12) [reply executor/aggregate_test.go]: Now I change the panic possible to `5%`, to reduce the retry times.

File: executor/aggregate.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/aggregate.go:297]: What's the `+1` for ?

Comment (wshwsh12) [reply executor/aggregate.go:297]: fetchChildData() also use the channel.

File: executor/aggregate.go [hunk 4]
```diff
@@ -736,13 +743,13 @@ func (e *HashAggExec) waitPartialWorkerAndCloseOutputChs(waitGroup *sync.WaitGro
...
-	close(e.finalOutputCh)
-}
-
 func (e *HashAggExec) prepare4ParallelExec(ctx context.Context) {
-	go e.fetchChildData(ctx)
+	fetchChildWorkerWaitGroup := &sync.WaitGroup{}
+	fetchChildWorkerWaitGroup.Add(1)
+	go func() {
+		e.fetchChildData(ctx)
+		fetchChildWorkerWaitGroup.Done()
```
Comment (XuHuaiyu) [inline executor/aggregate.go:759]: We can pass `fetchChildWorkerWaitGroup` as an input argument to `fetchChildData`.

File: executor/aggregate.go [hunk 5]
```diff
@@ -771,11 +778,20 @@ func (e *HashAggExec) prepare4ParallelExec(ctx context.Context) {
...
 	}
 	go func() {
-		e.waitFinalWorkerAndCloseFinalOutput(finalWorkerWaitGroup)
+		finalWorkerWaitGroup.Wait()
 		if e.stats != nil {
 			atomic.AddInt64(&e.stats.FinalWallTime, int64(time.Since(finalStart)))
 		}
 	}()
+
+	go func() {
```
Comment (XuHuaiyu) [inline executor/aggregate.go:787]: Extract a function like `waitAllWorkersAndCloseFinalOutput`

File: executor/aggregate.go [hunk 6]
```diff
@@ -784,14 +789,9 @@ func (e *HashAggExec) prepare4ParallelExec(ctx context.Context) {
...
 
-	go func() {
-		// All workers maybe send error message to e.finalOutputCh when they panic.
-		// And e.finalOutputCh should be closed after all goroutines gone.
-		fetchChildWorkerWaitGroup.Wait()
-		partialWorkerWaitGroup.Wait()
-		finalWorkerWaitGroup.Wait()
-		close(e.finalOutputCh)
-	}()
+	// All workers maybe send error message to e.finalOutputCh when they panic.
```
Comment (XuHuaiyu) [inline executor/aggregate.go:792]: maybe --> may

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**